dimod/include/dimod/constrained_quadratic_model.h

```diff
@@ -0,0 +1,493 @@
+    for (auto& c : constraints_) {
+        c.parent_ = this;
+    }
+}
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>::ConstrainedQuadraticModel(
+        ConstrainedQuadraticModel&& other) noexcept
+        : ConstrainedQuadraticModel() {
+    swap(*this, other);
+}
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>&
+ConstrainedQuadraticModel<bias_type, index_type>::operator=(ConstrainedQuadraticModel other) {
+    swap(*this, other);
+    return *this;
+}
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>&
+ConstrainedQuadraticModel<bias_type, index_type>::operator=(
+        ConstrainedQuadraticModel&& other) noexcept {
+    swap(*this, other);
+    return *this;
+}
+
+template <class bias_type, class index_type>
+index_type ConstrainedQuadraticModel<bias_type, index_type>::add_constraint() {
+    constraints_.emplace_back(this);
+    return constraints_.size() - 1;
+}
+
+template <class bias_type, class index_type>
+template <class B, class I, class T>
+index_type ConstrainedQuadraticModel<bias_type, index_type>::add_constraint(
+        const abc::QuadraticModelBase<B, I>& lhs, Sense sense, bias_type rhs,
+        const std::vector<T>& mapping) {
+    // todo: move version
+    assert(mapping.size() == objective.num_variables());
+
+    add_constraint();
+    Constraint<bias_type, index_type>& constraint = constraints_.back();
+
+    for (size_type i = 0; i < lhs.num_variables(); ++i) {
+        assert(vartype(mapping[i]) == lhs.vartype(i));
+        assert(lower_bound(mapping[i]) == lhs.lower_bound(i));
+        assert(upper_bound(mapping[i]) == lhs.upper_bound(i));
+
+        constraint.add_linear(mapping[i], lhs.linear(i));
+    }
+
+    // quadratic biases
+    for (auto it = lhs.cbegin_quadratic(); it != lhs.cend_quadratic(); ++it) {
+        constraint.add_quadratic(mapping[it->u], mapping[it->v], it->bias);
+    }
+
+    // offset
+    constraint.add_offset(lhs.offset());
+
+    // sense and rhs
+    constraint.set_sense(sense);
+    constraint.set_rhs(rhs);
+
+    return constraints_.size() - 1;
+}
+
+template <class bias_type, class index_type>
+index_type ConstrainedQuadraticModel<bias_type, index_type>::add_constraints(index_type n) {
+    assert(n >= 0);
+    index_type size = constraints_.size();
+    for (index_type i = 0; i < n; ++i) {
+        constraints_.emplace_back(this);
+    }
+    return size;
+}
+
+template <class bias_type, class index_type>
+index_type ConstrainedQuadraticModel<bias_type, index_type>::add_linear_constraint(
+        std::initializer_list<index_type> variables, std::initializer_list<bias_type> biases,
+        Sense sense, bias_type rhs) {
+    Constraint<bias_type, index_type> constraint(this);
+
+    auto vit = variables.begin();
+    auto bit = biases.begin();
+    for (; vit != variables.end() && bit != biases.end(); ++vit, ++bit) {
```

Comment:
What happens when one iterator ends up shorter than the other ?

Reply:
It uses the shortest.